File: ruby/lib/redis/bg_save.rb
require_relative '../redis/commands/save'

module Redis
  class BgSave
    SLEEP_INTERVAL = 5.freeze
    
    # @params [Hash]
    # keys: store [Hash], save_pid [Integer(pid)]
    def self.run(store:, save_pid:)
      Thread.new do
        loop do
          begin
            Process.getpgid(save_pid)
          rescue Errno::ESRCH
          rescue
            unless store.size.zero?
              puts "Saving keys: #{store.size}"
              Redis::Commands::Save.run(store: store)
            end
          end

          # sleep the thread for 5 seconds
          sleep(SLEEP_INTERVAL)
        end
      end
    end
  end
end

File: python/src/bg_save.py
from src.commands.save import Save
import os
import asyncio

SLEEP_INTERVAL = 5

async def bg_save(store: dict, saved_pid: int):
  while True:
    try:
      child_id, _ = os.waitpid(saved_pid, os.WNOHANG)
      if child_id == 0:
        pass
    except TypeError:
      print(f"Saving keys: {len(store)}")
      Save(store=store).run()
    
    await asyncio.sleep(SLEEP_INTERVAL)
